clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/handlers/LoadBalancerUpsertHandler.groovy

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright 2016 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.clouddriver.aws.deploy.handlers
+
+import com.amazonaws.AmazonServiceException
+import com.amazonaws.services.elasticloadbalancing.AmazonElasticLoadBalancing
+import com.amazonaws.services.elasticloadbalancing.model.ApplySecurityGroupsToLoadBalancerRequest
+import com.amazonaws.services.elasticloadbalancing.model.CreateLoadBalancerListenersRequest
+import com.amazonaws.services.elasticloadbalancing.model.CreateLoadBalancerRequest
+import com.amazonaws.services.elasticloadbalancing.model.DeleteLoadBalancerListenersRequest
+import com.amazonaws.services.elasticloadbalancing.model.Listener
+import com.amazonaws.services.elasticloadbalancing.model.LoadBalancerDescription
+import com.netflix.spinnaker.clouddriver.data.task.Task
+import com.netflix.spinnaker.clouddriver.data.task.TaskRepository
+import com.netflix.spinnaker.clouddriver.orchestration.AtomicOperationException
+
+class LoadBalancerUpsertHandler {
+
+  private static final String BASE_PHASE = "UPSERT_ELB"
+
+  private static Task getTask() {
+    TaskRepository.threadLocalTask.get()
+  }
+
+  public static void updateLoadBalancer(AmazonElasticLoadBalancing loadBalancing, LoadBalancerDescription loadBalancer,
+                                  List<Listener> listeners, Collection<String> securityGroups) {
+    def amazonErrors = []
+    def loadBalancerName = loadBalancer.loadBalancerName
+
+    if (securityGroups) {
+      loadBalancing.applySecurityGroupsToLoadBalancer(new ApplySecurityGroupsToLoadBalancerRequest(
+        loadBalancerName: loadBalancerName,
+        securityGroups: securityGroups
+      ))
+      task.updateStatus BASE_PHASE, "Security groups updated on ${loadBalancerName}."
+    }
+
+    if (listeners) {
+      // ignore all references to :0 => :0 listeners - leave them alone if they're there, do not add them if they're not
+      listeners = listeners.findAll(notLegacyListener);
+      def existingListeners = loadBalancer.listenerDescriptions*.listener.findAll(notLegacyListener)
```

Comment:
Addressed the double-filtering by moving it up - still needs to happen in both places but this is a little clearer.